crates/wasi-common/src/file.rs

```diff
@@ -1,67 +1,73 @@
 use crate::{Error, ErrorExt, SystemTimeSpec};
 use bitflags::bitflags;
 use std::any::Any;
+use std::sync::{Arc, RwLock};
+
+#[cfg(unix)]
+use cap_std::io_lifetimes::AsFd;
+#[cfg(windows)]
+use io_extras::os::windows::AsRawHandleOrSocket;
 
 #[wiggle::async_trait]
 pub trait WasiFile: Send + Sync {
     fn as_any(&self) -> &dyn Any;
-    async fn get_filetype(&mut self) -> Result<FileType, Error>;
+    async fn get_filetype(&self) -> Result<FileType, Error>;
 
     #[cfg(unix)]
-    fn pollable(&self) -> Option<rustix::fd::BorrowedFd> {
+    fn pollable(&self) -> Option<Arc<dyn AsFd + '_>> {
         None
     }
 
     #[cfg(windows)]
-    fn pollable(&self) -> Option<io_extras::os::windows::RawHandleOrSocket> {
+    fn pollable(&self) -> Option<Arc<dyn AsRawHandleOrSocket + '_>> {
         None
     }
```

Comment:
How come these need changing? The prior signatures I think should work?

Reply:
In the method a `RwLockWriteGuard<BorrowedFd>` would be produced, which can't be dereferenced for return.

Reply:
Sorry let me restate. From a system API perspective a lock is not needed here, so I don't know where the rwlock is motivated from. I would expect that these signatures remain the same since this is intended for bottoming out in the underlying platform-specific interface which handles all synchronization internally in the kernel.

Reply:
It's only because of `File::set_fd_flags()` needing &mut self